### ISSUE TITLE
Django ORM 2006 mysql gone away

### ISSUE THREAD

**Issue by ofpiyush** [encode/apistar#233]
The default timeout on mysql is 28800 seconds (8 hours).

After that, all requests fail till the server is restarted.

We need to find a good solution to this problem.

In the interim I have found a simple hack which relies on  the `request_started` signal django uses internally.

```python
from functools import wraps

from django.core.signals import request_started


def django_wsgi_hack(func):
    @wraps(func)
    def wrapper(environ, *args, **kwargs):
        request_started.send(sender=func.__class__, environ=environ)
        return func(environ, *args, **kwargs)
    return wrapper


app = App(...)
app.wsgi = django_wsgi_hack(app.wsgi)
```

**Comment by ssteinerx** [comment]
Django's default CONN_MAX_AGE is 0 (zero) which closes the connection at the end of each request:  https://docs.djangoproject.com/en/1.11/ref/settings/#conn-max-age

Which specific MySQL connection timeout setting(s) are involved?

**Comment by ofpiyush** [comment]
@ssteinerx `wait_timeout` was the one I believe I ran up against.

[Django's WSGIHandler](https://github.com/django/django/blob/cecc07/django/core/handlers/wsgi.py#L150-L152) uses [the `request_started` signal](https://github.com/django/django/blob/cecc079/django/db/__init__.py#L59-L65) to [close a connection](https://github.com/django/django/blob/cecc07/django/db/backends/base/base.py#L523-L525) based off [the `CONN_MAX_AGE` setting](https://github.com/django/django/blob/cecc07/django/db/backends/base/base.py#L183-L184) you're talking about.

**Comment by ssteinerx** [comment]
So, are you suggesting that DjangoBackend should use CONN_MAX_AGE (or some other, similar setting) to manage the connection to the database?

**Comment by ofpiyush** [comment]
I am bringing up a problem I faced and have provided a solution that I am using right now.

I don't know how APIStar wishes to solve the problem.

For example, 

One could just close db connections by hand in DjangoBackend's build. 

Or call the func this signal calls.

Or one could override a connection's execute method and handle similar db errors there.

Or use the hack I'm using here to let other listeners of the signal do their work. Eg: reset query cache.

I don't have an answer, definitely not one that suits the larger community, especially when half the community doesn't even use Django's ORM.

But I'd really like for the conversation to happen. 😊

**Comment by tomchristie** [comment]
0.2 is going to bring in support for components as context managers, which will make things easier to resolve as we'll have enter/exit on the ORM component. (Also in due course we'll likely be working towards deployment options for running API Star within either of Flask or Django completely)

**Comment by tomchristie** [comment]
This oughta be resolved now, using `Session` in 0.2 onwards.
